csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -489,114 +412,244 @@ std::unique_ptr<ReductionParams> innerOuterPersistentHeuristic(
+  // elements in the inner dimension after vectorization.
+  // Start from 128 or a smaller number if inner dim is small.
+  const int64_t after_vect = inner_dim_numel / vect_factor;
+  const int64_t batch_min = getMinimumBatch();
+  int64_t threads_per_block_min =
+      InnerOuterPersistentKernelScheduler::threads_per_block_min;
+  threads_per_block_min = std::min(threads_per_block_min, after_vect);
+  threads_per_block_min = scheduler_utils::roundUpPow2(threads_per_block_min);
+
+  // star max threads per block from min threads per block
+  int64_t threads_per_block_max = threads_per_block_min;
+  // increase to cover the whole inner dim
+  threads_per_block_max =
+      std::max(threads_per_block_max, ceilDiv(after_vect, batch_min));
+  // round up to power of 2
+  threads_per_block_max = scheduler_utils::roundUpPow2(threads_per_block_max);
+  // don't go beyond the maximum threads per block
+  threads_per_block_max = std::min(
+      threads_per_block_max,
+      InnerOuterPersistentKernelScheduler::threads_per_block_max);
+
+  // Store all the possible heuristics based on different threads per block.
+  // Vectorizaton is fixed at the maximum value.
+  std::vector<InnerOuterParams> iop_candidates;
+  for (auto threads_per_block = threads_per_block_max;
+       threads_per_block >= threads_per_block_min;
+       threads_per_block /= 2) {
+    iop_candidates.emplace_back(
+        getHeuristicsGivenVectThreads(vect_factor, threads_per_block));
   }
-  // Step-4, set OuterParams Reduction dim: bdimy.
-  iop.bdimy = threads_per_block / iop.bdimx;
-  NVF_ERROR(
-      iop.bdimy * iop.bdimx == threads_per_block,
-      " threads_per_block must be divisible by bdimx and bdimy.");
-  // Step-5, special case, when inner_dim_numel <= 1024, bdimx is usually small
+
+  // Sort the heuristics based on the register usage and occupancy.
+  std::stable_sort(
+      iop_candidates.begin(),
+      iop_candidates.end(),
+      [](const InnerOuterParams& a, const InnerOuterParams& b) {
+        // If a thread can use more registers than required, there is a high
+        // chance that it can void register spilling and compiler can optimize
+        // for better instruction level parallelism.
+        int64_t extra_regs_a =
+            a.available_register_per_thread - a.required_register_per_thread;
+        int64_t extra_regs_b =
+            b.available_register_per_thread - b.required_register_per_thread;
+        if (extra_regs_a > 0 && extra_regs_b < 0) {
+          return true;
+        } else if (extra_regs_a < 0 && extra_regs_b > 0) {
+          return false;
+        }
+        // High occupancy provides better threads level parallelism.
+        // 25% is sufficient since ILP is high due to persistent batch sizes
+        // which is equivalent to unrolling inner dim.
+        if (a.warps_per_sm < 16 || b.warps_per_sm < 16) {
+          return a.warps_per_sm > b.warps_per_sm;
+        }
```

Comment:
should we add `a.warps_per_sm != b.warps_per_sm` in the condition so we would try to compare the cta size then?

Reply:
good point, added.